### PR TITLE
Make HF converge with lighting native amp

### PR DESCRIPTION
**Pull request by pengwa**
#### Problem:

Huggingface Transformers: 4.22.0
PyTorch Lightning: 1.6.3 
PyTorch: v1.12.1, cuda 11.6
ORT: main branch, cuda 11.6

Model: RobertaForSequenceClassification @ models/roberta/modeling_roberta.py
Mixed Precision training with `torch.autocast`: https://github.com/Lightning-AI/lightning/blob/a64e1dfd7d2be51887a151d893623cd0570041bd/pytorch_lightning/plugins/precision/native_amp.py#L99

Under this amp autocast context, forward + loss computation run. Here is a snippet of loss computation.

```
        if labels is not None:
                ...
            if self.config.problem_type == "regression":
                loss_fct = MSELoss()
                if self.num_labels == 1:
                   ...
            elif self.config.problem_type == "single_label_classification":
                loss_fct = CrossEntropyLoss()
                **loss = loss_fct(logits.view(-1, self.num_labels), labels.view(-1))**
            elif self.config.problem_type == "multi_label_classification":
                ...

        return SequenceClassifierOutput(
            loss=loss,
            logits=logits,
            hidden_states=outputs.hidden_states,
            attentions=outputs.attentions,
        )
```

It is found after forward run, loss is 1.0850 in float16, looks good.. Then it did a scaling up here: https://github.com/Lightning-AI/lightning/blob/a64e1dfd7d2be51887a151d893623cd0570041bd/pytorch_lightning/plugins/precision/native_amp.py#L62, the scaler is 65536. then we get a scaled loss 71104 in float type (because float16 loss multiple fp32 scaler, type got promoted to fp32). Then backward started with initial grads to be 1, then 1 (float32) * 65536 (float32) as the backward step, generating a float16 gradient, then we got a `inf`. The problem occurs. With `inf`, the backward feed the `inf` into crossentropygradient op, generating `nan`s. Then all gradients got `nan` in back propagation. 

So we see training with ORTModule (it almost always `overflow`, the loss did not drop too much, as compared with PyTorch).

#### Analysis for the UT (when autocast enabled)

PyTorch trace graph looks like this :

```
graph(%0 : Float(16, 3, strides=[3, 1], requires_grad=0, device=cuda:0),
      %target : Long(16, strides=[1], requires_grad=0, device=cuda:0),
      %2 : Float(3, 3, strides=[3, 1], requires_grad=1, device=cuda:0)):
  %9 : int = prim::Constant[value=5]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %10 : bool = prim::Constant[value=0]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %11 : bool = prim::Constant[value=0]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %12 : NoneType = prim::Constant()
  %13 : Half(3, 3, strides=[3, 1], requires_grad=0, device=cuda:0) = aten::to(%2, %9, %10, %11, %12) # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %14 : int = prim::Constant[value=5]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %15 : bool = prim::Constant[value=0]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %16 : bool = prim::Constant[value=0]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %17 : NoneType = prim::Constant()
  %18 : Half(16, 3, strides=[3, 1], requires_grad=0, device=cuda:0) = aten::to(%0, %14, %15, %16, %17) # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %19 : NoneType = prim::Constant()
  %input : Half(16, 3, strides=[3, 1], requires_grad=0, device=cuda:0) = aten::linear(%18, %13, %19) # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
  %21 : NoneType = prim::Constant()
  %22 : int = prim::Constant[value=1]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/functional.py:3,014:0
  %23 : int = prim::Constant[value=-100]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/functional.py:3,014:0
  %24 : float = prim::Constant[value=0.]() # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/functional.py:3,014:0
  %data : Float(requires_grad=0, device=cuda:0) = **aten::cross_entropy_loss(%input, %target, %21, %22, %23, %24) # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/functional.py:3,014:0**
  %27 : Float(requires_grad=0, device=cuda:0) = ^_OutputIdentityOp()(%data) # /opt/conda/envs/ptca/lib/python3.8/site-packages/onnxruntime/training/ortmodule/_io.py:430:0
  return (%27)
```

The most important lines 

%target : Long(16, strides=[1], requires_grad=0, device=cuda:0),
%input : **_Half_**(16, 3, strides=[3, 1], requires_grad=0, device=cuda:0) = aten::linear(%18, %13, %19) # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/modules/linear.py:114:0
**_Float_**(requires_grad=0, device=cuda:0) = aten::cross_entropy_loss(**%_input_**, %target, %21, %22, %23, %24) # /opt/conda/envs/ptca/lib/python3.8/site-packages/torch/nn/functional.py:3,014:0

`aten::cross_entropy_loss` takes Half input, and return Float output. As said in doc: https://pytorch.org/docs/stable/amp.html#cuda-ops-that-can-autocast-to-float32, `cross_entropy` in autocast mode will run in fp32 mode, e.g. convert its input to fp32 (if it is not), do the compute and return fp32 result. The other hand, ORT's `SoftmaxCrossEntropyLossInternal` take same types of input and output, and our code https://github.com/microsoft/onnxruntime/blob/31cb3cb254615a1414e7323a4da9ff0b1a0f727d/orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py#L68 when exporting `aten::cross_entropy_loss` assumed this, and set the output to be fp16 either. So this is the reason we have the problem.

#### Possible Fixes
1. Enhance `SoftmaxCrossEntropyLossInternal` to support different types of input and output. 
2. Check the input and output when exporting, add the input case explicitly if there is type promotion from input to output.

This PR used the 2nd approach. We can start 1st approach when needed later. 

TODO: revisit all other exporter functions, add the checks, etc.

### CODE REVIEW

**Comment by iK1D** [comment]
Any reason we don't follow the approach 1 but chose 2? It's common that loss function cannot be compute on FP16, during ORTTrainer, we once had transformer to force loss sub-graph to be computed on float. I am thinking the custom op registration from our side is just a hack to exporter, ideally we don't need them and just let exporter to handle (not sure if the ATEN_FALLBACK mode works or not from exporter team for now), not sure if it's a good way to add python code for this single Op work from frontend but make the code logic more complex.

**Comment by pengwa** [comment]
> Any reason we don't follow the approach 1 but chose 2? It's common that loss function cannot be compute on FP16, during ORTTrainer, we once had transformer to force loss sub-graph to be computed on float. I am thinking the custom op registration from our side is just a hack to exporter, ideally we don't need them and just let exporter to handle (not sure if the ATEN_FALLBACK mode works or not from exporter team for now), not sure if it's a good way to add python code for this single Op work from frontend but make the code logic more complex.

Approach 1 requires tons of changes (I was working in local branch), it may take some time to be in after reviewed. 

As long as the "the custom op registration from our side is just a hack to exporter" is needed, I did not see much complexity that is introduced by this change. Most of the logic remains, and the only thing added ruing exporting is we add a cast. This is the extra complexity that can be removed once Approach 1 is done.

**Comment by pengwa** [comment]
Oh if you mean the added input output data type checks, I think we need add them for safety, failing fast will be better than hiding precision issue for us to find issue and get a fix.

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py [hunk 1]
```diff
@@ -12,6 +14,86 @@
 from onnxruntime.training import ortmodule
 
 
+def _enforce_check(condition, message):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py:17]: Is this code movement required? Can we keep this file only for custom symbolic registration, and move all helper functions to another utility/helper file?

**Review by baijumeswani — DISMISSED** [review]
Just a small comment. But otherwise looks good.

File: orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py (hunk 1 above)
Comment (pengwa) [reply orttraining/orttraining/python/training/ortmodule/_custom_op_symbolic_registry.py]: Yeah, let me clean up the code in it and avoid adding more things than the fix only needs.

My original intention is, PythonOp is one of the operators to export, so maybe we should move some of the shared custom export logic into _custom_op_symbolic_registry.py. While looks we treat this file as a workaround collection against exporter defect. If so, I would move those changes unrelated to the fix to its original places, and just leave the fix here. If possible, those fix should also be up streamed to torch exporter later.  FYI @iK1D

**Review by baijumeswani — APPROVED** [review]

**Review by askhade — APPROVED** [review]
LGTM
